sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/ConnectionHandler.java

```diff
@@ -0,0 +1,157 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.io.rabbitmq;
+
+import com.rabbitmq.client.Channel;
+import com.rabbitmq.client.Connection;
+import com.rabbitmq.client.ConnectionFactory;
+import com.rabbitmq.client.MissedHeartbeatException;
+import com.rabbitmq.client.PossibleAuthenticationFailureException;
+import com.rabbitmq.client.ShutdownSignalException;
+import java.io.Closeable;
+import java.io.IOException;
+import java.net.ProtocolException;
+import java.net.URISyntaxException;
+import java.security.KeyManagementException;
+import java.security.NoSuchAlgorithmException;
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.TimeoutException;
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * RabbitMQ multiplexes over a single Connection using Channels so it should not be necessary to
+ * open multiple Connections to a single host. This class implements {@link ChannelLeaser} by
+ * maintaining a map of lessees and their associated Channels over a single Connection.
+ *
+ * <p>Note: it is unclear what, if anything, is ultimately responsible for calling close() here; it
+ * should only happen on Beam runner shutdown, or if it were knowable that there were no current
+ * Readers or Writers interacting with Rabbit.
+ */
+@ThreadSafe
+class ConnectionHandler implements ChannelLeaser, Closeable {
+  private final Map<UUID, Channel> channelsByLessee = new ConcurrentHashMap<>();
+  private final String uri;
+  private volatile Connection connection;
+
+  public ConnectionHandler(String uri) {
+    this.uri = uri;
+  }
+
+  @Override
+  public <T> T useChannel(UUID lesseeId, ChannelLeaser.UseChannelFunction<T> f) throws IOException {
+    try {
+      Channel channel = getChannel(lesseeId);
+      return f.apply(channel);
+    } catch (PossibleAuthenticationFailureException
+        | ProtocolException
+        | MissedHeartbeatException e) {
+      // full Connection-level problem
+      close();
+      throw e;
+    } catch (ShutdownSignalException e) {
+      // Connection- or Channel-level, depending
+      Object cause = e.getReference();
+      if (cause instanceof Channel) {
+        closeChannel(lesseeId);
+      }
+      if (cause instanceof Connection) {
+        close();
+      }
+      throw e;
+    }
+  }
+
+  @Override
+  public void closeChannel(UUID lessee) {
+    Channel toClose = channelsByLessee.remove(lessee);
+    if (toClose != null) {
+      try {
+        toClose.close();
+      } catch (IOException | TimeoutException e) {
+        // ignore
```

Comment:
On the off chance it did, it would ultimately be closed as part of closing the Connection after this. But almost assuredly if this fails, the connection/channel have already been severed.